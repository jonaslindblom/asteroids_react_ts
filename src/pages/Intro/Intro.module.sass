.container
  display: flex
  flex-direction: column
  justify-content: flex-start
  height: 100vh
  overflow-y: auto
  overflow-x: hidden
  background-color: #2f4f4f
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='80' height='105' viewBox='0 0 80 105'%3E%3Cg fill-rule='evenodd'%3E%3Cg id='death-star' fill='%239C92AC' fill-opacity='0.12'%3E%3Cpath d='M20 10a5 5 0 0 1 10 0v50a5 5 0 0 1-10 0V10zm15 35a5 5 0 0 1 10 0v50a5 5 0 0 1-10 0V45zM20 75a5 5 0 0 1 10 0v20a5 5 0 0 1-10 0V75zm30-65a5 5 0 0 1 10 0v50a5 5 0 0 1-10 0V10zm0 65a5 5 0 0 1 10 0v20a5 5 0 0 1-10 0V75zM35 10a5 5 0 0 1 10 0v20a5 5 0 0 1-10 0V10zM5 45a5 5 0 0 1 10 0v50a5 5 0 0 1-10 0V45zm0-35a5 5 0 0 1 10 0v20a5 5 0 0 1-10 0V10zm60 35a5 5 0 0 1 10 0v50a5 5 0 0 1-10 0V45zm0-35a5 5 0 0 1 10 0v20a5 5 0 0 1-10 0V10z' /%3E%3C/g%3E%3C/g%3E%3C/svg%3E")
  background-image: url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' width='24' height='24' viewBox='0 0 24 24'%3E%3Cg fill='%239C92AC' fill-opacity='0.12'%3E%3Cpolygon fill-rule='evenodd' points='8 4 12 6 8 8 6 12 4 8 0 6 4 4 6 0 8 4'/%3E%3C/g%3E%3C/svg%3E")

  @media (min-height: 600px)
    justify-content: space-around

  .topButtons
    position: absolute
    top: 0
    right: 0

  .infoButton
    background-color: tomato
    color: white
    font-size: clamp(16px, 4vmin, 30px)
    height: clamp(30px, 6vmin, 40px)
    width: clamp(30px, 6vmin, 40px)
    border-radius: 50%
    cursor: pointer
    text-transform: uppercase
    margin: 3vmin
    padding: 0

    @media (hover: hover) and (pointer: fine)
    &:hover
      background-color: brown

  .content
    display: flex
    flex-direction: column
    align-items: center
    justify-content: space-evenly
    color: tomato
    min-width: 320px
    height: 100%
    min-height: 568px
    overflow: hidden

    .version
      position: absolute
      left: 0
      top: 0
      font-size: clamp(10px, 2vmin, 20px)
      font-weight: bolder
      color: gray
      margin: 3vmin

    .logo
      margin-top: 48px
      display: flex
      flex-direction: column
      align-items: center

      .prefix
        font-size: 6vmin
        font-family: "Pacifico", cursive
        text-shadow: 1px 1px 10px #fff
        color: lightgray
        transform: translate(-100%, 30%) rotate(-13deg)

      .title
        font-size: clamp(3rem, 12vmin, 10rem)
        text-shadow: 1px 1px 10px #fff, 1px 1px 10px #ccc
        margin: 8px
        margin-top: 8px

    .suffix
      font-size: 4vmin
      font-family: "Pacifico", cursive
      text-shadow: 1px 1px 10px #fff
      color: lightgray
      text-align: center

    .ship
      width: 25vmin
      height: 25vmin
      max-width: 200px
      display: flex
      justify-content: center
      align-items: center
      animation: spin 10s linear infinite

      @keyframes spin
        0%
          transform: rotate(0deg)
        100%
          transform: rotate(360deg)

    .playButtons
      display: flex
      flex-wrap: nowrap
      gap: 8px
      justify-content: center
      align-items: center
      width: 100%

      .playButton
        font-size: 3vmin
        font-weight: bolder
        cursor: pointer

      @media (hover: hover) and (pointer: fine)
        button:hover
          background-color: brown

    .highscore
      text-align: center
      font-size: clamp(12px, 4vmin, 20px)
      margin: 1vmin
      color: gray

    .bottomButtons
      font-size: 0.75rem
      display: flex
      flex-direction: row
      justify-content: center
      align-items: center
      margin: 1vmin
      gap: 12px

    .bottomSpace
      width: 100%
      height: 42px
