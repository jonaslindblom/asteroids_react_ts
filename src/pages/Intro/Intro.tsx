import styles from "./Intro.module.sass";
import { useState, useRef } from "react";
import InfoView from "./components/InfoView";
import { ReactComponent as SpaceShip } from "src/pages/Scene/components/Ship/assets/ship.svg";
import { useLocalStorage } from "src/hooks/useLocalStorage";
import Button from "src/components/Button";
import { useNavigate, Link } from "react-router-dom";

export default function Intro() {
  const [showInfo, setShowInfo] = useState(false);
  const [highScore] = useLocalStorage("highscore", 0);

  const intro = useRef(null);
  const navigate = useNavigate();

  return (
    <div className="fadeIn">
      <div className={styles.container}>
        {!showInfo ? (
          <div className={styles.topButtons}>
            <button
              className={styles.infoButton}
              onClick={() => {
                setShowInfo(true);
              }}
              data-testid="info-button"
            >
              ?
            </button>
          </div>
        ) : undefined}
        <div className={styles.content} ref={intro}>
          {intro.current && showInfo && (
            <InfoView
              container={intro.current}
              onClose={() => {
                setShowInfo(false);
              }}
            />
          )}
          <div className={styles.version} data-testid="version">
            v. {process.env.VERSION}
          </div>
          <div className={styles.logo}>
            <div className={styles.prefix}>Incoming</div>
            <div className={styles.title}>
              <strong>Asteroids</strong>
            </div>
            <div className={styles.suffix}>from Jonas&apos; deep space</div>
          </div>
          <div className={styles.ship}>
            <SpaceShip />
          </div>
          <div className={styles.playButtons}>
            <Button
              classes={styles.playButton}
              onClick={() => {
                navigate("/play");
              }}
              data-testid="play-button"
            >
              Let&rsquo;s play!
            </Button>
          </div>
          <div className={styles.highscore}>Device high score: {highScore}</div>
          <div className={styles.bottomButtons}>
            <Link to="/attribution">Attribution</Link>
            <Link to="/privacy">Privacy Policy</Link>
          </div>
          <div className={styles.bottomSpace}></div>
        </div>
      </div>
    </div>
  );
}
