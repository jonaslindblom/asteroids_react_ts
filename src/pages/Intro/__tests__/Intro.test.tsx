import { render, fireEvent } from "@testing-library/react";
import Intro from "../Intro";
import { MemoryRouter } from "react-router-dom";
import { vi } from "vitest";
import * as router from "react-router";

const navigate = vi.fn();

function removeElementByTestId(element: HTMLElement, testId: string) {
  const target = element.querySelector(`[data-testid="${testId}"]`);
  if (target) {
    target.remove();
  }
}

describe("Intro", () => {
  beforeEach(() => {
    vi.spyOn(router, "useNavigate").mockImplementation(() => navigate);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it("renders component", () => {
    const utils = render(
      <MemoryRouter>
        <Intro />
      </MemoryRouter>
    );
    const score = utils.getByText("Asteroids");
    expect(score).toBeInTheDocument();
  });

  it("navigates to /play when start button is clicked", () => {
    const utils = render(
      <MemoryRouter>
        <Intro />
      </MemoryRouter>
    );
    const startButton = utils.getByTestId("play-button");
    expect(startButton).toBeInTheDocument();
    fireEvent.click(startButton);
    expect(navigate).toHaveBeenCalledWith("/play");
  });

  it("opens info popup", () => {
    const utils = render(
      <MemoryRouter>
        <Intro />
      </MemoryRouter>
    );
    const infoButton = utils.getByTestId("info-button");
    expect(infoButton).toBeInTheDocument();
    fireEvent.click(infoButton);
    expect(navigate).not.toHaveBeenCalled();
    const about = utils.getByText("About");
    expect(about).toBeInTheDocument();
    const qrCode = utils.getByTestId("qr-code");
    expect(qrCode).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Intro />
      </MemoryRouter>
    );
    removeElementByTestId(container, "version"); // Version is automatically bumped, so remove from snapshot
    expect(container).toMatchSnapshot();
  });
});
