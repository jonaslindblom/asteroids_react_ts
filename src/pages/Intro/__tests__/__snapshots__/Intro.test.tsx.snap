// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Intro > matches snapshot 1`] = `
<div>
  <div
    class="fadeIn"
  >
    <div
      class="_container_add5c9"
    >
      <div
        class="_topButtons_add5c9"
      >
        <button
          class="_infoButton_add5c9"
          data-testid="info-button"
        >
          ?
        </button>
      </div>
      <div
        class="_content_add5c9"
      >
        <div
          class="_logo_add5c9"
        >
          <div
            class="_prefix_add5c9"
          >
            Incoming
          </div>
          <div
            class="_title_add5c9"
          >
            <strong>
              Asteroids
            </strong>
          </div>
          <div
            class="_suffix_add5c9"
          >
            from Jonas' deep space
          </div>
        </div>
        <div
          class="_ship_add5c9"
        >
          <svg
            height="800px"
            viewBox="0 0 64 64"
            width="800px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m17.86 29c0-.21 2.83-8.84 2.83-8.84s-.6-4.9-.36-5a5.24 5.24 0 0 1 1.09-.07s-.34-2.91 0-3.05a4.66 4.66 0 0 1 2-.1c0 .17-.09 1.38 0 1.3s3.2-7.6 5.66-8.92 5.62-1.05 6.92 1.38 3.39 7.65 3.39 7.65 0-1.54.16-1.59 2.38.06 2.43.35-.05 2.58-.05 2.58 1.68.33 1.73.66-.66 7.19-.66 7.19 2.24 4.1 2.36 5.31-2.2 8.15-3.18 11.64-2.28 7.44-2.28 7.44l2.52-.86a14.16 14.16 0 0 1 .87-3 23.66 23.66 0 0 0 2.43-2.28c1.38-1.37 1.49-1.92 1.49-1.92s-.68-7.52-.39-7.74a3.43 3.43 0 0 1 1.18-.32s-.26-1.74 0-1.91a9.48 9.48 0 0 1 3 .16c0 .13.37-1.47.79-1.52s2.21.46 3.58 4.06 1.49 12.4-.08 16.84-3.72 6.9-7.44 7.89-6.15.76-6.78.57-1.22-.38-1.19-.84.2-1.67.2-1.67h-2.79a12.85 12.85 0 0 1 -.89 2.32 4.5 4.5 0 0 1 -1.42.12 16.51 16.51 0 0 1 -.36 3.35c-.24.17-5.95.46-6.16.14a30.56 30.56 0 0 1 -.66-3.78s-1.54.09-1.63-.12-1-2.89-1-2.89l-3-.16a11.45 11.45 0 0 1 .42 3c-.19.46-4.93.89-8.56-.67s-5.77-3.25-6.3-10-.17-14.38.53-17.36 1.49-3.25 1.95-3.18a.91.91 0 0 1 .77.73c0 .29.12 1.08.12 1.08s2.45-.33 2.5.09a12.34 12.34 0 0 1 -.07 1.71s1.7-.18 1.77.62.07 9.71.07 9.71a20 20 0 0 0 2.79 2.66c.76.4.67.11.76.4s1 4.47 1.26 4.63a34.23 34.23 0 0 0 3.44.48s-5.02-13.09-5.76-18.27z"
              fill="#1d1d1b"
            />
            <path
              d="m32 4.94c2.33 0 3.41 2.57 5 6.23s6.54 16.77 6.57 17.9-8 26-8.24 26-.49.06-.5-.15 2.63-12.58 2.56-13.58-3.84-13.44-4.32-14.01-3.23-.61-3.63-.22-2.84 15.38-2.74 16 2.87 11.34 2.54 11.43a1.94 1.94 0 0 1 -1.16.12c-.34-.16-7.69-23.45-7.91-25.28s2.8-10.88 4.6-14.85 4.06-9.62 7.23-9.59z"
              fill="#e6e4da"
            />
            <path
              d="m27.6 12.9c.31-.47 2.22-1.23 4.55-1.21a8 8 0 0 1 4.16 1c.09.16 2.82 7.92 2.45 8.1s-3.75-1.18-7.33-.95-6.38 1.1-6.6.94 2.17-6.94 2.77-7.88z"
              fill="#1d1d1b"
            />
            <path
              d="m32.1 12.65a7.23 7.23 0 0 1 3.57.85c.26.33 2 5.74 1.67 5.79a21.19 21.19 0 0 0 -6.28-.86c-3.2.22-4.8 1.27-4.81 1s1.83-6 2.16-6.14a9.8 9.8 0 0 1 3.69-.64z"
              fill="#cbe7f5"
            />
            <path
              d="m28.6 10c.08-.25 1.53-3.3 1.91-3.35s.78-.32.54.11-1.41 3.08-1.65 3.24-.8 0-.8 0z"
              fill="#1d1d1b"
            />
            <path
              d="m23.69 22.81s.95 0 .92.19-1.61 4.59-1.82 4.8-.62.27-.75.11 1.65-5.1 1.65-5.1z"
              fill="#1d1d1b"
            />
            <path
              d="m22.47 31c0-.3.66-.23.83.06s1.21 4.21 1 4.3-.57.44-.71 0-1.12-4.13-1.12-4.36z"
              fill="#1d1d1b"
            />
            <path
              d="m30.9 28.69c.14-.17.92 0 1 .14s2.35 7.72 2.18 7.73-1.4-.92-1.4-.8-.26.85-.09 1 1.78 1.19 1.92 1.61.79 2.64.79 2.64a19.85 19.85 0 0 0 -2.19-2c-.12.09-.36.43-.19.59s2.23 2.06 2.12 2.44a3.84 3.84 0 0 0 -.15.63 18.23 18.23 0 0 0 -2.25-1.48c-.13.09-.37.47-.15.68s2.21 1.51 2.22 1.76-.21 1.43-.21 1.43a8.85 8.85 0 0 0 -1.81-.57c0 .12-.3 1-.05 1.16s1.53.84 1.53.84-1.34 8.33-1.59 8.38-1.37.21-1.5 0-2.68-11.64-2.65-12 2.23-13.87 2.47-14.18z"
              fill="#e7413e"
            />
            <g
              fill="#1d1d1b"
            >
              <path
                d="m39.28 25.77a9.83 9.83 0 0 1 2.35 2 1.36 1.36 0 0 1 -.28.55s-2.55-1.76-2.6-1.93.49-.45.53-.62z"
              />
              <path
                d="m37.59 27.78c0-.28 3.54 2.94 3.55 3.11s0 .58-.19.5a35.14 35.14 0 0 1 -3.95-2.88c-.07-.29.55-.51.59-.73z"
              />
              <path
                d="m36.82 29.89a23.72 23.72 0 0 1 3.94 3.55c.05.29.05.29-.12.38a21.49 21.49 0 0 1 -4.14-3.3c-.05-.37.2-.46.32-.63z"
              />
              <path
                d="m36.85 32.64a25 25 0 0 1 3.47 3.07c0 .16.1.37-.07.42a34.12 34.12 0 0 1 -3.84-2.89c-.04-.17.36-.6.44-.6z"
              />
              <path
                d="m37.15 35.51a26.89 26.89 0 0 1 2.52 2.09c0 .13 0 .71-.15.67a18.86 18.86 0 0 1 -2.73-2.17c0-.16.36-.59.36-.59z"
              />
            </g>
            <path
              d="m48.1 32.27c.18-.1 1.91-.27 1.87-.1a46.66 46.66 0 0 1 -1.85 5.06 39.17 39.17 0 0 1 -.02-4.96z"
              fill="#f2910d"
            />
            <path
              d="m49.08 30.36s-.34-.36.07-.46a6.7 6.7 0 0 1 1.38 0s0 .67-.19.72a6.12 6.12 0 0 1 -1.26-.26z"
              fill="#f2910d"
            />
            <path
              d="m45 44.11a31.14 31.14 0 0 0 5.13-6.9c1.85-3.76 2.4-6.24 2.61-6.29s1.32 2.46 1.38 5.92.24 8.08-.71 11.07a11.35 11.35 0 0 1 -3.36 4.89s-2.63-4.42-2.88-4.41-.75.1-.62.35 2.85 4.58 2.68 4.75a6.89 6.89 0 0 1 -1.23.51s-2-3.86-2.24-3.85-.87.23-.78.36 2.22 3.59 2.02 3.76a1.32 1.32 0 0 1 -.57.22l-1.79-2.61s-.66.23-.53.43 1.53 2.25 1.24 2.38-3 .55-3 .26 2.65-10.67 2.65-10.84z"
              fill="#a42116"
            />
            <path
              d="m48.44 46.76a20.87 20.87 0 0 1 2.65 3.68c0 .17-.48.55-.65.48a42.41 42.41 0 0 1 -2.53-3.64c-.09-.28.49-.39.53-.52z"
              fill="#1d1d1b"
            />
            <path
              d="m49.48 45.36c0-.21.45-.43.66-.4a18.87 18.87 0 0 1 2.26 3.11c0 .2.14.66-.24.5a22 22 0 0 1 -2.68-3.21z"
              fill="#1d1d1b"
            />
            <path
              d="m50.72 43.61s.4-.72.53-.6 2.06 2.11 2 2.4 0 .71-.23.64-2.17-2.32-2.3-2.44z"
              fill="#1d1d1b"
            />
            <path
              d="m51.82 41.29c0-.13.06-.75.31-.76s1.24 1 1.33 1.25.07.87-.05.84-1.59-1.33-1.59-1.33z"
              fill="#1d1d1b"
            />
            <path
              d="m29 56.8a45.8 45.8 0 0 1 4.75-.1c0 .17 0 2.63-.34 2.72a22.27 22.27 0 0 1 -4 .08c-.07-.17-.64-2.69-.41-2.7z"
              fill="#f2910d"
            />
            <path
              d="m11.53 27.72c.15-.16 1.17-.12 1.21 0s.16 1.12 0 1.21-1.33.25-1.34.08a10.88 10.88 0 0 1 .13-1.29z"
              fill="#f2910d"
            />
            <path
              d="m11.48 30.09c0-.29 2.37-.07 2.38.1s.44 6.57.23 6.45a20.54 20.54 0 0 1 -2.61-6.55z"
              fill="#f2910d"
            />
            <path
              d="m9.86 31.85c0-.39.87 5.48 3.19 7.87a22.16 22.16 0 0 0 4.23 3.5s2.58 8.14 2.41 8-1.78-2.41-2-2.45-.57.39-.53.52 2.73 3.46 2.86 3.75.41 1.15.29 1.24-.41.18-.5 0a43.65 43.65 0 0 0 -3.64-4.52c-.33-.08-.58 0-.66.18s3.47 4.49 3.22 4.49a3.88 3.88 0 0 1 -1.25-.09c-.17-.12-2.57-3.51-2.73-3.58s-.83.35-.7.56 2.51 2.88 2.26 2.89a7 7 0 0 1 -5.71-4.17c-1.6-3.74-.9-16.48-.74-18.19z"
              fill="#a42116"
            />
          </svg>
        </div>
        <div
          class="_playButtons_add5c9"
        >
          <div
            class="_container_89d844"
          >
            <button
              class="_playButton_add5c9 _button_89d844 "
              data-testid="play-button"
              type="button"
            >
              Letâ€™s play!
            </button>
          </div>
        </div>
        <div
          class="_highscore_add5c9"
        >
          Device high score: 
          0
        </div>
        <div
          class="_bottomButtons_add5c9"
        >
          <a
            href="/attribution"
          >
            Attribution
          </a>
          <a
            href="/privacy"
          >
            Privacy Policy
          </a>
        </div>
        <div
          class="_bottomSpace_add5c9"
        />
      </div>
    </div>
  </div>
</div>
`;
